pkg/sqlcache/db/client.go

```diff
@@ -29,8 +30,59 @@ const (
 	informerObjectCachePerms fs.FileMode = 0o600
 )
 
-// Client is a database client that provides encrypting, decrypting, and database resetting.
-type Client struct {
+// Client is a database client that provides encrypting, decrypting, and database resetting
+type Client interface {
+	WithTransaction(ctx context.Context, forWriting bool, f WithTransactionFunction) error
+	Prepare(stmt string) *sql.Stmt
+	QueryForRows(ctx context.Context, stmt transaction.Stmt, params ...any) (*sql.Rows, error)
+	ReadObjects(rows Rows, typ reflect.Type, shouldDecrypt bool) ([]any, error)
+	ReadStrings(rows Rows) ([]string, error)
+	ReadInt(rows Rows) (int, error)
+	Upsert(tx transaction.Client, stmt *sql.Stmt, key string, obj any, shouldEncrypt bool) error
+	CloseStmt(closable Closable) error
+	NewConnection() error
+}
+
+// WithTransaction runs f within a transaction.
+//
+// If forWriting is true, this method blocks until all other concurrent forWriting
+// transactions have either committed or rolled back.
+// If forWriting is false, it is assumed the returned transaction will exclusively
+// be used for DQL (e.g. SELECT) queries.
+// Not respecting the above rule might result in transactions failing with unexpected
+// SQLITE_BUSY (5) errors (aka "Runtime error: database is locked").
+// See discussion in https://github.com/rancher/lasso/pull/98 for details
+//
+// The transaction is committed if f returns nil, otherwise it is rolled back.
+func (c *client) WithTransaction(ctx context.Context, forWriting bool, f WithTransactionFunction) error {
+	c.connLock.RLock()
+	// note: this assumes _txlock=immediate in the connection string, see NewConnection
+	tx, err := c.conn.BeginTx(ctx, &sql.TxOptions{
+		ReadOnly: !forWriting,
+	})
+	c.connLock.RUnlock()
+	if err != nil {
+		return err
+	}
+
+	err = f(transaction.NewClient(tx))
+
+	if err != nil {
+		rerr := tx.Rollback()
+		err = errors.Join(err, rerr)
```

Comment:
Shouldn't we be appending `rerr` and `cerr` only if they're non-nil?

Reply:
[Join discards nils](https://pkg.go.dev/errors#Join), so my understanding is the current code should be correct.